rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+readable, but they result in the same outcome.
+
+The object under iteration is not mutated, instead a copy of the value is
+iterated, and mutated, returning a new object or array after iteration
+completes.
+
+#### Array Iteration
+
+The signature for array iteration is as follows:
+
+```coffee
+map(value: ARRAY, recursive: BOOLEAN) |<index variable>, <value variable>| EXPRESSION -> ARRAY
+```
+
+This is nearly identical to the object signature, except that it takes an array,
+has a variable for the index of the current item, and returns a single
+expression to use as the value at the given index.
+
+Here's an example:
+
+```json
+["foo", "bar"]
+```
+
+```coffee
+. = map(.) |index, value| {
+    value + "_" + to_string!(index)
+}
+```
+
+```json
+["foo_0", "bar_1"]
+```
+
+### Implementation
+
+This proposal favors adding a _mapping_ function over _for-loop syntax_. That
+is, the RFC proposes:
+
+```coffee
+map(.) |key, value| [upcase(key), value]
+```
+
+over:
+
+```coffee
+for (key, _value) in . {
+  key = upcase(key)
+}
+```
+
+This choice is made both on technical merits, based on the [VRL Design
+Document][doc] and for improved future capabilities. See the
+"[for-loop](#for-loop)" alternative section for more details on this.
+
+For the chosen proposal to work, there are two separate concepts that need to
+be implemented:
+
+- closure-support for functions
+- limited lexical scoping
+
+Let's discuss these one by one, before we arrive at the final part, implementing
+the `map` function that uses both concepts.
+
+[doc]: https://github.com/timberio/vector/blob/jean/vrl-design-doc/lib/vrl/DESIGN.md
+
+#### Closure-support For Functions
+
+For iteration to land in the form proposed in this RFC, we need a way for
+operators to write _what_ they want to happen to keys and/or values of objects
+and arrays.
+
+We do this by allowing functions to expose the fact that they accept a closure
+as a stand-alone argument to their function call.
+
+"stand-alone" means the closure comes _after_ the function call itself, e.g.
+this:
+
+```coffee
+map(.) |k, v| {
+  [k, v]
+}
+```
+
+over this:
+
+```coffee
+map(., |k, v| {
+  [k, v]
+})
+```
+
+This choice is made to make it clear that closures in VRL _can't be passed
+around through variables, but are instead syntactically attached to a function
+call_.
+
+That is, we don't want to allow this:
+
+```coffee
+my_closure = |k, v| [k, v]
+map(., my_closure)
+```
+
+As that would require us to (1) allow for a closure to be represented as a `Value`,
+since everything assigned to a variable has to be representable as a value, and
+(2) it would require us to add compiler-support to reject assigning closures to
+an event field (e.g. `.foo = |k, v| [k, v]`), as that would lead to its own set
+of problems.
+
+While closures-as-values might seem useful at first, there hasn't been any
+use-case that has come up in the past that makes this a requirement.
```

Comment:
On the other hand, if we'd allow closures in a variable any special handling for recursion (the `recursive: true` argument or `recursive for` construct) would not be needed at the language level.

```
upcase_keys = |key, value| {
  [upcase(key), if is_object(value) {
    upcase_keys(value)
  } else {
    value
  }]
}

map(., upcase_keys)
```

I briefly talked about it with @JeanMertz and he made the case that the `recursive` constructs in this RFC might be more easily understood by users that are not very familiar with programming languages - which I agree with.

Interested in what the others think about this.

Reply:
The argument you attributed to me is correct, and harks back to  the "**[design the feature for the intended target audience](https://github.com/timberio/vector/blob/jean/vrl-design-doc/lib/vrl/DESIGN.md#target-audience)**" section of the VRL design document.

There are more (more pressing) reasons why I'm against allowing this, though.

The most important one, is that allowing this can result in poor-performing code, even resulting in infinite-loops. The most important rule in the VRL design document states **[Safety and performance over ease of use](https://github.com/timberio/vector/blob/jean/vrl-design-doc/lib/vrl/DESIGN.md#the-zen-of-vrl)**. Which this technique would violate.

Secondly, I want to keep variable and target assignment (e.g. `foo = true` and `.foo = true`) semantically equal in VRL. You can assign any value to a variable that you can also assign to an event field, and vice-versa.

When you assign a closure to a variable, you are assigning _code that is optionally evaluated at runtime_. We cannot do this for event field assignment without seeing a significant increase in cognitive complexity on assignment rules, or by rejecting assigning closures to event fields, which again makes this a non-starter to me.

Reply:
I'll add the above comments to the RFC, to better explain why this was rejected as a viable option.

Reply:
Added note in da1b44a71bea17e2c883960565a1aafa8f7a7fc1.